server/src/main/scala/com/cloudera/livy/server/recovery/SessionStore.scala

```diff
@@ -69,14 +65,17 @@ class SessionStore(
 
   /**
    * Return the next unused session id with specified session type.
-   * If checks the SessionManagerState stored and returns the next free session id.
+   * It checks the SessionManagerState stored and returns the next free session id.
    * If no SessionManagerState is stored, it returns 0.
+    * It saves the new session ID to the session store.
    *
    * @throws Exception If SessionManagerState stored is corrupted, it throws an error.
    */
-  def getNextSessionId(sessionType: String): Int = {
-    store.get[SessionManagerState](sessionManagerPath(sessionType))
+  def getNextSessionId(sessionType: String): Int = synchronized {
+    val nextSessionId = store.get[SessionManagerState](sessionManagerPath(sessionType))
```

Comment:
`SessionStore` calls into the distributed Atomic recipe from Apache curator.
This is the how I am doing it:

```scala
  val distributedSessionId = new DistributedAtomicLong(curatorClient, zkPath, retryPolicy)

  def nextBatchSessionId: Int = {
    recursiveTry(distributedSessionId, MAX_RETRY) match {
      case Some(sessionId) =>
        sessionId.toInt
      case None =>
        val msg: String = "Failed to get the next session id from Zookeeper"
        logger.warn(msg)
        throw new IOException(msg)
    }
  }

  @tailrec
  private def recursiveTry(distributedLong: DistributedLong, retryCount: Int): Option[Long] = {
    val updatedValue = distributedLong.increment
    updatedValue.succeeded match {
      case _ if retryCount <= 0 =>
        None
      case true if retryCount > 0 =>
        Option(updatedValue.preValue())
      case _ =>
        recursiveTry(distributedLong, retryCount - 1)
    }
  }
```